### ISSUE TITLE
Can I run necessist with nextest?

### ISSUE THREAD

**Issue by moodmosaic** [trailofbits/necessist#935]
*What a great tool* 👍 💯

When using `necessist` in Rust, is it agnostic when it comes to the underlying test runner? I'd be interested to run this on a project where tests are run via `cargo nextest run`.

**Comment by smoelius** [comment]
> _What a great tool_ 👍 💯

Thanks. :)

> When using necessist in Rust, is it agnostic when it comes to the underlying test runner?

Currently, it is hardcoded to use `cargo test`: https://github.com/trailofbits/necessist/blob/cade4663886c35b59cf64cedb488541ef93c0000/frameworks/src/rust/mod.rs#L550-L551

Does your project require `cargo nextest`?

**Comment by moodmosaic** [comment]
It'd be preferred to use `cargo nextest`. I can also use `cargo test` but have to pass some args, for example:

```
cargo test testnet  -- --test-threads=1
```

If there's a way to pass arguments to `cargo test` using `necessist` then it'd be good enough for me to start.

**Comment by smoelius** [comment]
I think this should work:
```
necessist -- -- --test-threads=1
```
The first `--` is to pass arguments to `cargo test`. The second is to pass arguments to the test executable.

Aside: I'm not opposed to using alternate test runners. But my understanding is that the main benefit to using `cargo nextest` is to run test executables in parallel. Currently, Necessist runs test executables one-at-a-time, so I can't immediately see a benefit to using `cargo nextest`. By I might be convinced otherwise.

**Comment by smoelius** [comment]
To filter which tests are run, please try passing a list of test files to Necessist, e.g.:
```
necessist path-to-file-containing-pox_4_tests ... -- <cargo test args> -- <test executable args>
```
This interface is a consequence of wanting to support multiple languages. But I now see that it can be confusing.

I'm not sure what the best "fix" is. Maybe a warning could be issued if the additional arguments match some pattern (e.g., contain `--package`).

**Comment by moodmosaic** [comment]
Thanks! That worked! Thank you for all the help. This issue can be closed now.

---

Given all the above, there are errors/warnings when running the tool against a fairly sized codebase ([stacks-core](https://github.com/stacks-network/stacks-core)), and I am happy to open a separate issue if you think I should.

```
necessist ./stackslib/src/chainstate/stacks/boot/pox_4_tests.rs \
-- --package stackslib -- chainstate::stacks::boot::pox_4_tests

229 candidates in 1 test file

stackslib/src/chainstate/stacks/boot/pox_4_tests.rs: dry running
stackslib/src/chainstate/stacks/boot/pox_4_tests.rs: mutilating
stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:187:5-187:53:
  Warning: Failed to run test `chainstate::stacks::boot::pox_4_tests::pox_extend_transition`
           This may indicate a bug in Necessist. Consider opening an issue at:
           https://github.com/trailofbits/necessist/issues

stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:215:5-216:71:
 Warning: Failed to run test
stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:383:9-383:67:
 Warning: Failed to run test
stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:416:5-416:63:
 Warning: Failed to run test
stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:431:5-431:63:
 Warning: Failed to run test
stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:441:9-441:67:
 Warning: Failed to run test
stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:445:5-445:48:
 Warning: Failed to run test
stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:501:5-504:7:
 Warning: Failed to run test
stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:510:9-510:45:
 Warning: Failed to run test
stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:627:20-627:32:
 Warning: Failed to run test
stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:693:5-693:65:
 Warning: Failed to run test
stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:781:5-781:53:
 Warning: Failed to run test
stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:797:13-797:47:
 Warning: Failed to run test
stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:801:5-804:7:
 Warning: Failed to run test
stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:831:5-831:38:
 Warning: Failed to run test
stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:840:5-843:7:
 Warning: Failed to run test
stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:857:9-857:75:
 Warning: Failed to run test
stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:862:9-862:64:
 Warning: Failed to run test
stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:869:21-869:52:
 Warning: Failed to run test

████████████████████████████████████████████░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░ 124/229
Error: Ctrl-C detected

nikos@ubuntu-linux:~/stacks-core$

```

**Comment by smoelius** [comment]
> Given all the above, there are errors/warnings when running the tool against a fairly sized codebase ([stacks-core](https://github.com/stacks-network/stacks-core)), and I am happy to open a separate issue if you think I should.

Is the code publicly available so that I might try to reproduce the issue?

If not, do you experience similar behavior with some other code (e.g., in the linked repository)?

**Comment by moodmosaic** [comment]
>Is the code publicly available so that I might try to reproduce the issue?

Thank you for asking. Of course, here's how you can reproduce it:

```bash
# Install package dependencies (skip if you already have them):
sudo apt install build-essential libclang-dev pkg-config libssl-dev

# Install Rust (skip if you already have it):
curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh
source "$HOME/.cargo/env"

# Clone stacks-core, cd into it, switch to `next` branch:
git clone https://github.com/stacks-network/stacks-core.git && cd stacks-core
git checkout next

# Run `pox_4_tests` via `cargo test`:
cargo nextest run chainstate::stacks::boot::pox_4_tests

# Run `pox_4_tests` via `necessist`:
necessist ./stackslib/src/chainstate/stacks/boot/pox_4_tests.rs \
-- --package stackslib -- chainstate::stacks::boot::pox_4_tests
```

**Comment by smoelius** [comment]
I think this command may do what you want:
```sh
necessist stackslib/src/chainstate/stacks/boot/pox_4_tests.rs \
-- --package stackslib chainstate::stacks::boot::pox_4_tests
#                     ^ No `--` here.
```

I gave you bad advice. The double `--` doesn't work. The reason is that Necessist [passes `-- --exact <testname>`](https://github.com/trailofbits/necessist/blob/e05123887a2068fb31f0a527920c17eefe43013e/frameworks/src/rust/mod.rs#L537) at the end of the test command. If the extra arguments contain a `--`, then the `--` before `--exact` becomes the third in the command, not the second. The end result is an ill-formed test command.

Fortunately, `cargo test` allows `<testname>` to be passed before the `--`, which is why the above command works.

Aside: `RUST_LOG=debug` shows the commands that are executed, if you're interested. That's how I debugged this.

**Comment by smoelius** [comment]
> Thank you! That command executed fine.

I'm really glad to hear that.

> Now, how do you I interpret the output of `necessist` after it's finished running? In the output below, each `passed` occurrence points to a potential bug in the test(?)

All the output tells you is that if you remove those statements/method calls, the tests still pass.

So giving a direct answer to your question ("each `passed` occurrence points to a potential bug in the test(?)") is difficult.

I suspect that these are false positives and can be ruled out immediately;
```
stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:501:5-504:7: `info!(
        "Block height: {}",
        get_tip(peer.sortdb.as_ref()).block_height
    );` passed
```
```
stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:797:13-797:47: `assert_latest_was_burn(&mut peer);` passed
```
In other words, it's not surprising for a test to still pass if it prints less information (the former) or performs fewer assertions (the latter).

This one looks more suspicious:
```
stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:1156:9-1156:71: `latest_block = peer.tenure_with_txs(&[], &mut coinbase_nonce);` passed
```
It seems to suggest that the `latest_block` variable need not be updated. That could be a bug, or it could not. One would need to scrutinize the test to tell.

BTW, you might consider createing a necessist.toml file with the following content:
```toml
ignored_functions = ["assert_latest_was_burn"]
ignored_macros = ["info"]
```
Doing so will eliminate many of those likely false positives, and will make Necessist finish sooner.
